addons/hr_expense/models/hr_expense.py

```diff
@@ -150,8 +151,30 @@ def _compute_state(self):
     @api.depends('quantity', 'unit_amount', 'tax_ids', 'currency_id')
     def _compute_amount(self):
         for expense in self:
+            if expense.product_id and not expense.product_has_cost:
+                continue
             taxes = expense.tax_ids.compute_all(expense.unit_amount, expense.currency_id, expense.quantity, expense.product_id, expense.employee_id.user_id.partner_id)
-            expense.total_amount = taxes.get('total_included')
+            expense.total_amount = taxes['total_included']
+
+    @api.depends('total_amount', 'tax_ids', 'currency_id')
+    def _compute_amount_tax(self):
+        for expense in self:
+            taxes = expense.tax_ids.compute_all(price_unit=expense.total_amount, currency=expense.currency_id, quantity=1.0, product=expense.product_id, partner=expense.employee_id.user_id.partner_id)
+            expense.amount_tax = taxes['total_included'] - taxes['total_excluded'] if expense.tax_ids else 0.0
+            expense.untaxed_amount = taxes['total_excluded']
```

Comment:
This looks wrong.
You are computing the amounts for only 1 unit.
These computations need to be done in the `_compute_amount` method.

Reply:
Indeed.
Then the fields in company currency can be computed in the same function too. No need to duplicate these functions.
`_compute_amount_tax_company` and `_compute_total_amount_company` are almost copy pasted, and the discussion on this PR proved that it is not a good idea to copy paste because a fix applied at one place might not be applied to the "same" function right below it.


_______________________________
Another way of doing it could be:
*However, I'm not sure if it is a good idea because of the stored/non stored mix in the depends*
```python
currency_rate = fields.Float(compute=_'compute_currency_rate')

@api.depends('date', 'currency_id', 'company_currency_id')
def _compute_currency_rate(self):
    # only get the conversion rate once, it can be a relatively expensive part 
    # compared to the rest of the computation
    for expense in self:  
        expense.currency_rate = self.env['res.currency']._get_conversion_rate(
            from_amount=1, 
            from_currency=expense.currency_id,
            to_currency=expense.company_currency_id,
            company=expense.company_id, 
            date=date_expense,
        )

@api.depends('currency_rate', 'amount_tax')
def _compute_amount_tax_company(self):
    for expense in self:
        expense.amount_tax_company = expense.amount_tax * expense.currency_rate

@api.depends('currency_rate', 'total_amount')
def _compute_amount_tax_company(self):  # still a bit of copy paste, but way less
    for expense in self:
        expense.total_amount_company = expense.total_amount * expense.currency_rate

@api.depends('currency_rate'')
def _compute_label_convert_rate(self):
    for expense in self:
        expense.label_convert_rate = _(
            '1 %(exp_cur)s = %(rate)s %(comp_cur)s', 
            exp_cur=expense.currency_id.name, 
            rate=float_repr(expense.currency_rate, expense.company_currency_id.decimal_places),     
            comp_cur=expense.company_currency_id.name,
        )
```

Reply:
By the way, using the company currency's precision to compute a rate is questionable too. 
For instance, 6 digits are used on the currency's form.

Let's say you convert from Rupiah to USD.
1 Rupiah = 0,000070 USD. But rounded to 2 digits because we take the USD precision.
Now 1 IDR = 0,00 USD :thinking:

Reply:
One thing at a time :sweat_smile: 

So, to start with the first comment first. We (@KangOl and I) discussed it and `_compute_amount` and `_compute_amount_tax` cannot be merged into one function and here is why.

Because there are cases, when total_amount can be set directly from the form and then untaxed_amount won't be computed (in case we put it inside  _compute_amount and for a reminder _compute_amount  has following dependency `@api.depends('quantity', 'unit_amount', 'tax_ids', 'currency_id')` ). That's why I think I have to keep two compute methods for total_amount and for untaxed_amount. There are lots of computed fields on expense that depend on each other. That does not make things easy, but to cover all the cases, that's the approach for me. Also adding product_id into dependency won't do a trick. We could have the same product and still change only total_amount from form and untaxed_amount should be computed.

So, we came to conclusion to add a comment that explains why we need `_compute_amount_tax` and to make a little helper method for the computation of the taxes with the right argument. 

Now I will check your suggestion @william-andre and be back to you ^^

Reply:
Soooo, indeed, having computed field for `currency_rate` is a great idea and I did made changes according to the suggestion. Thanks! :raised_hands: